### PR TITLE
Update all pipeline statuses in case failure

### LINKED ISSUE #25: Wait for all pipeline statuses are updated and change flow status after
In some cases, some of the pipelines may fail before others. Currently, once one of them is failed we are not updating statuses for rest - leaves us unclear how did other pipelines go. Eg we need status for validation report pipeline, but if the creation of csv fails first we will not ever get.

## Acceptance Criteria

* [ ] we know exact statuses for all pipelines 

## Tasks

* [ ] update tests
* [ ] refacro logic

## analysis

* Right now if any of the pipelines were failed, right away the flow status was changed to FAILED.
  * So if at least one failed pipeline in DB -> flow = FAILED -> delete pipelines
*  I changed this logic so that even if there is failed pipeline i'm waiting to get statuses for all other pipelines and changing flow_status after.
  *So now we have: if at least one pipeline is QUEUED or RUNNING wait -> no QUEUED and RUNNING and at least one failed -> flow = FAILED -> delete pipelines
Problem is that I'm never getting the update request for "main" pipeline if failed. Eg. take a look at following pseudo code

```
# pipeline ids
csv = me/mydata/1/derived_csv
report = me/mydata/1/derived_report
main = me/mydata/1

# dpp sends requests
specstore.update(csv, event=running, success=None )
specstore.update(csv, event=finish, success=False )

specstore.update(report, event=running, success=None )
specstore.update(report, event=finish, success=True )

# for some reason below is not happening when csv is fails
specstore.update(main, event=running, success=None)
specstore.update(main, event=finish, success=False)
```
the main one is always queued - not even running
Note this is not happening if successful - request for "main" pipeline is sent and updated fine

### PR DESCRIPTION
**Pull request by zelima**
* Wait for all pipeline statuses are updated and change flow status after
* Set status for depended pipelines to Failed if failed pipeline comes in
* simplify logic for getting flow status
* tests

Fixes #25

### CODE REVIEW

File: flowmanager/models.py [hunk 1]
```diff
@@ -266,24 +273,27 @@ def list_pipelines(self):
...
             success = session.query(Pipelines).filter_by(
                 flow_id=flow_id, status=STATE_SUCCESS).first()
             pending = session.query(Pipelines).filter_by(
                 flow_id=flow_id, status=STATE_PENDING).first()
+            # at least one pending and success -> Flow in Progress
             if (pending is not None) and (success is not None):
                 return STATE_RUNNING
-
-            # If non of success, failed or running status is queued/pending
+            failed = session.query(Pipelines).filter_by(
```
Comment (akariv) [inline flowmanager/models.py]: we never return `STATE_SUCCESS` here?

File: flowmanager/models.py [hunk 2]
```diff
@@ -266,24 +273,27 @@ def list_pipelines(self):
...
+                flow_id=flow_id, status=STATE_FAILED).first()
+            # at least one pending and failed -> Flow in Progress
+            if (pending is not None) and (failed is not None):
+                return STATE_RUNNING
+            # at least one failed and no pending or running -> Flow Failed
+            if failed is not None:
+                return STATE_FAILED
+            # If non of success, failed or running  -> Flow queued
             if (failed is None) and (success is None) and (running is None):
```
Comment (akariv) [inline flowmanager/models.py]: no need to test for `(running is None)` here

File: flowmanager/models.py [hunk 3]
```diff
@@ -266,24 +273,27 @@ def list_pipelines(self):
 
     def check_flow_status(self, flow_id):
```
Comment (akariv) [inline flowmanager/models.py]: The logic in this function is hard to follow, and as you can see also prone to bugs.

Let's try to rewrite in a more understandable way:
If at least one pipeline is running -> RUNNING

Otherwise, we have 7 combinations of (PENDING, SUCCESS, FAILED), we can mark them as:
(P, S, F, PS, PF, SF, PSF)

The mapping between each of the cases to flow states would be:
   1. all * PENDING -> PENDING    [P]
   2. some * PENDING and some * (FAILED / SUCCESS) -> RUNNING     [PF, PS, PSF]
   3. no PENDING and all * SUCCESS -> SUCCESS        [S]
   4. no PENDING and some * (FAILED / SUCCESS) -> FAILED     [F, FS]

You can see we covered all 7 cases.

So, we could write this as this pseudocode
```
if exists pending:
  if exists success or exists failed:
     -> running
  else:
     -> pending
else:
  if exists failed:
      -> failed
  else:
      -> success
```

File: flowmanager/models.py (hunk 1 above)
Comment (akariv) [reply flowmanager/models.py]: (probably means there's also a missing test, no?)

Comment (zelima) [reply flowmanager/models.py]: We return it at the end (if none of the others is satisfied). We do have tests for it
